src/bms/parse/validity.rs

```diff
@@ -0,0 +1,465 @@
+            }
+        }
+
+        // 1.2) Check BGMs reference valid WAV ids.
+        for obj_ids in self.notes.bgms.values() {
+            for obj_id in obj_ids {
+                if !self.notes.wav_files.contains_key(obj_id) {
+                    missing.push(ValidityMissing::WavForBgm(*obj_id));
+                }
+            }
+        }
+
+        // 1.3) Check BGAs reference valid BMP ids.
+        for bga_obj in self.graphics.bga_changes().values() {
+            if !self.graphics.bmp_files.contains_key(&bga_obj.id) {
+                missing.push(ValidityMissing::BmpForBga(bga_obj.id));
+            }
+        }
+
+        // 1.4) Check BPM change ids used in messages have corresponding #BPMxx definitions.
+        for id in &self.arrangers.bpm_change_ids_used {
+            if !self.scope_defines.bpm_defs.contains_key(id) {
+                missing.push(ValidityMissing::BpmChangeDef(*id));
+            }
+        }
+
+        // 1.5) Check STOP ids used in messages have corresponding #STOPxx definitions.
+        for id in &self.arrangers.stop_ids_used {
+            if !self.scope_defines.stop_defs.contains_key(id) {
+                missing.push(ValidityMissing::StopDef(*id));
+            }
+        }
+
+        // 3) Placement/overlap checks for notes on lanes.
+        //      - Visible notes in section 000
+        //      - Overlap: visible single vs single (same time, same lane)
+        //      - Overlap: visible single within long interval (same lane)
+        //      - Overlap: landmine within long interval -> warn once at long start
+        use std::collections::HashMap as StdHashMap;
+        let mut lane_to_notes: StdHashMap<(PlayerSide, Key), Vec<&Obj>> = StdHashMap::new();
+        for notes in self.notes.objs.values() {
+            for obj in notes {
+                // Visible note in section 000 (track index 0)
+                if obj.kind == NoteKind::Visible && obj.offset.track.0 == 0 {
+                    invalid.push(ValidityInvalid::VisibleNoteInTrackZero {
+                        side: obj.side,
+                        key: obj.key,
+                        time: obj.offset,
+                    });
+                }
+                lane_to_notes
+                    .entry((obj.side, obj.key))
+                    .or_default()
+                    .push(obj);
+            }
+        }
+        for ((side, key), objs) in lane_to_notes.into_iter() {
+            if objs.is_empty() {
+                continue;
+            }
+            // Sort by time
+            let mut lane_objs = objs;
+            lane_objs.sort_unstable_by_key(|o| o.offset);
+
+            // Build LN intervals by pairing consecutive Long notes
+            let mut long_times: Vec<ObjTime> = lane_objs
+                .iter()
+                .filter(|o| o.kind == NoteKind::Long)
+                .map(|o| o.offset)
+                .collect();
+            long_times.sort_unstable();
```

Comment:
```suggestion
```
It is needless to sort because `lane_objs` has already sorted and its order has unchanged.